EIPS/eip-225.md

```diff
@@ -0,0 +1,474 @@
+
+### Attack vector: Censoring signer
+
+Another interesting attack vector is if a signer (or group of signers) attempts to censor out blocks that vote on removing them from the authorization list. To work around this, we restrict the allowed minting frequency of signers to 1 out of N/2. This ensures that malicious signers need to control at least 51% of signing accounts, at which case it's game over anyway.
+
+### Attack vector: Spamming signer
+
+A final small attack vector is that of malicious signers injecting new vote proposals inside every block they mint. Since nodes need to tally up all votes to create the actual list of authorized signers, they need to track all votes through time. Without placing a limit on the vote window, this could grow slowly, yet unbounded. The solution is to place a ~~moving~~ window of W blocks after which votes are considered stale. ~~A sane window might be 1-2 epochs.~~ We'll call this an epoch.
+
+### Attack vector: Concurrent blocks
+
+If the number of authorized signers are N, and we allow each signer to mint 1 block out of K, then at any point in time N-K+1 miners are allowed to mint. To avoid these racing for blocks, every signer would add a small random "offset" to the time it releases a new block. This ensures that small forks are rare, but occasionally still happen (as on the main net). If a signer is caught abusing it's authority and causing chaos, it can be voted out.
+
+## Specification
+
+We define the following constants:
+
+ * **`EPOCH_LENGTH`**: Number of blocks after which to checkpoint and reset the pending votes.
+   * Suggested `30000` for the testnet to remain analogous to the mainnet `ethash` epoch.
+ * **`BLOCK_PERIOD`**: Minimum difference between two consecutive block's timestamps.
+   * Suggested `15s` for the testnet to remain analogous to the mainnet `ethash` target.
+ * **`EXTRA_VANITY`**: Fixed number of extra-data prefix bytes reserved for signer *vanity*.
+   * Suggested `32 bytes` to retain the current extra-data allowance and/or use.
+ * **`EXTRA_SEAL`**: Fixed number of extra-data suffix bytes reserved for signer seal.
+   * `65 bytes` fixed as signatures are based on the standard `secp256k1` curve.
+ * **`NONCE_AUTH`**: Magic nonce number `0xffffffffffffffff` to vote on adding a new signer.
+ * **`NONCE_DROP`**: Magic nonce number `0x0000000000000000` to vote on removing a signer.
+ * **`UNCLE_HASH`**: Always `Keccak256(RLP([]))` as uncles are meaningless outside of PoW.
+ * **`DIFF_NOTURN`**: Block score (difficulty) for blocks containing out-of-turn signatures.
+   * Suggested `1` since it just needs to be an arbitrary baseline constant.
+ * **`DIFF_INTURN`**: Block score (difficulty) for blocks containing in-turn signatures.
+   * Suggested `2` to show a slight preference over out-of-turn signatures.
```

Comment:
We discussed this on Discord already, I would strongly suggest having the `DIFF_INTURN` set to `3`. 

Short summary: because it is more likely that you have two ouf-of-turn blocks with total score of 2, than three out-of-turn blocks with total score of 3. An in-turn block can help the network to reorganize even if there were two out-of-turn blocks. We had multiple of such occasions that caused Görli network to halt.